### DISCUSSION TITLE
Array state update with push method return none

### DISCUSSION THREAD

**Discussion by DrLoiseau** [typst/typst#3081] [Q&A] [answered]
### Description

Hi,

I'm encountering problems when trying to update a state of type array. If I update it by joining elements with the + operator it works fine, but if I use the push method, the new state is always none. I'd prefer to make an array of arrays instead of one long array. Here is the code I'm using :
```
#import "@preview/tablex:0.0.7": gridx

#let q_array = state("q_array", ())

#let question_count = counter("question")

#let question(question_text, bareme: none, comp: none) = {
    q_array.display()
    question_count.step()
    q_array.update(arr => arr + (bareme, comp))
    gridx(
        columns: (95%, 5%),
        inset: 0pt,
        align: (left, right),
        [*Question #question_count.display() :* #question_text ],
        [ /#bareme]
    )
    q_array.display()
}
```
This works fine, but when I change the function to update q_array state to `arr => arr.push((bareme, comp))` my state gets updated to `none` type. I tried to assign the new array to a variable and to return this variable but I get the same result. It doesn't seem like a feature to me and I haven't seen anything in the docs about array methods not been usable in state update. Do someone understand what's happening here ? Did I miss something ?

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by EpicEricEE** [comment, marked as answer]
The function in `update` needs to return the updated state, but `push` returns `none`. Also, the `+` operator concatenates arrays, that's why you end up with one long one. You have two options:

```typ
#q_array.update(arr => arr + ((bareme, comp),))
```

```typ
#q_array.update(arr => {
  arr.push((bareme, comp))
  arr
})
```
